subsys/bluetooth/audio/bap_broadcast_source.c

```diff
@@ -643,13 +643,64 @@ static enum bt_bap_ep_state broadcast_source_get_state(struct bt_bap_broadcast_s
 	return stream->ep->status.state;
 }
 
+static bool merge_bis_and_subgroup_data_cb(struct bt_data *data, void *user_data)
+{
+	struct bt_audio_codec_cfg *codec_cfg = user_data;
+	int err;
+
+	err = bt_audio_codec_cfg_set_val(codec_cfg, data->type, data->data, data->data_len);
+	if (err < 0) {
+		LOG_DBG("Failed to set type %u with len %u in codec_cfg: %d", data->type,
+			data->data_len, err);
+
+		return false;
+	}
+	return true;
+}
+
+static int update_codec_cfg_data(struct bt_audio_codec_cfg *codec_cfg,
+				 const struct bt_bap_broadcast_source_stream_param *stream_param)
+{
+	int err;
+	/* Merge subgroup codec configuration with the BIS configuration
+	 * As per the BAP spec, if a value exist at level 2 (subgroup) and 3 (BIS), then it is
+	 * the value at level 3 that shall be used
+	 */
+	if (codec_cfg->id == BT_HCI_CODING_FORMAT_LC3) {
+		err = bt_audio_data_parse(stream_param->data, stream_param->data_len,
+					  merge_bis_and_subgroup_data_cb, codec_cfg);
+		if (err != 0) {
+			LOG_DBG("Could not merge BIS and subgroup config in codec_cfg: %d", err);
+
+			return -EINVAL;
+		}
+	} else {
+		/* If it is not LC3, then we don't know how to merge the subgroup and BIS codecs,
+		 * so we just append them
+		 */
+		if (codec_cfg->data_len + stream_param->data_len >
+		    sizeof(codec_cfg->data)) {
+			LOG_DBG("Could not store BIS and subgroup config in codec_cfg (%u > %u)",
+				codec_cfg->data_len + stream_param->data_len,
+				sizeof(codec_cfg->data));
+
+			return -ENOMEM;
+		}
+
+		memcpy(&codec_cfg->data[codec_cfg->data_len], stream_param->data,
+		       stream_param->data_len);
```

Comment:
I see that this is also a bug in the broadcast sink: We are neither here (or in the broadcast sink) increasing `codec_cfg->data_len` after appending them